Rude/Rude.cs

```diff
@@ -8,208 +8,77 @@
-                }
-                else
-                {
-                    var actorOutRudes = actorOutRudeQuery.Count();
-                    var medianQuery =
-                            from r in Rudes
-                            where r.Timestamp > oneYear
-                            group Math.Log((r.Timestamp - oneYear).TotalMinutes) by r.ActorId into g
-                            select g.Sum();
-                    var median = medianQuery.Median();
-
-                    double durationFactor = actorOutRudes == 0 ? 2 : median.Value / actorOutRudes;
-                    durationFactor = Math.Max(durationFactor, 0.25);
-                    durationFactor = Math.Min(durationFactor, 2);
-                    durationFactor *= rng.NextDouble() * targetInRudes;
-                    var duration = TimeSpan.FromHours(durationFactor * 2).Ticks;
-                    rudeEntity.Duration = TimeSpan.FromTicks(Math.Min(duration, TimeSpan.FromDays(1).Ticks));
-                }
-
-                // add rude
-                Rudes.Add(rudeEntity);
-            }
+            rude.RaiseRudeLevel();
 
-            if (kick)
+            if (rude.RudeLevel > Prefixes.Length)
             {
-                await target.Kick(REASONS[rng.Next(REASONS.Length)]);
+                RudeStatus.Remove(actor.UserId);
+                await actor.Kick("You have been too rude.");
             }
             else
             {
-                target.Name = GetUsername(target.Name, targetInRudes, GetReudigLevel(target.UserId));
-                await target.SaveChanges();
+                actor.Name = GetUserName(actor);
+                await actor.SaveChanges();
             }
-            await actor.SendMessage("Ruded with: " + rudeEntity.Duration);
         }
 
         public override async Task OnUserConnected(IUser user)
         {
-            user.Name = GetUsername(user.UserId, user.Name);
+            user.Name = GetUserName(user);
             await user.SaveChanges();
         }
 
-        [Command("rude")]
-        public async Task RudeCommand(IUser user)
-        {
-            var inRudes = GetActiveInRudes(user.UserId);
-            var inKicks = GetKickRudes(user.UserId, inRudes);
-            var mapped = inRudes.Select((r, i) => "Rude #" + i + ": " + r.Duration);
-            mapped = mapped.Concat(inKicks.Select((r, i) => "SelfRude #" + i + ": " + r.Duration));
 
-            await user.SendMessage(String.Join("<br>", mapped));
-        }
 
-        private string GetUsername(int userId, string name)
+        private void EnsureRudeStatus(int userId)
         {
-            var inRudes = GetActiveInRudes(userId).Count();
-            var reudigLevel = GetReudigLevel(userId);
-            return GetUsername(name, inRudes, reudigLevel);
+            if (!RudeStatus.ContainsKey(userId))
+                RudeStatus[userId] = new RudeStatus
+                {
+                    OnRudeLevelDecrease = async () =>
```

Comment:
I don't like this. You're never actually using this abstraction - you're just using it as a crutch for your data structures. Design them properly (add userid and rudeplugin to rudestatus) and you can drop this entirely.